Top recoil hook cleaned, backport to CMSSW_13_2_X

vslokenb opened the pull request:
#### PR description:

Backport of recoilToTop. Pushed from 13_2_0_pre2.

#### PR validation:

Same as linked original PR below. Verified using `scram build code-check` and `scram build code-format`. Seem to be a couple of minor indentation changes in lines unrelated to the actual TopRecoilHook in Pythia8Hadronizer.cc as a result of the code-format tool.  All differences between original PR are to account for CMSSW_13_2_X formatting needs.

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

Original PR found [here](https://github.com/cms-sw/cmssw/pull/42180). Backport requested by maintainers.

Commit:
Backport to 13_2

Commit:
Rest of backport to 13_2

cmsbuild commented:
A new Pull Request was created by  @vslokenb  for CMSSW_13_2_X.

It involves the following packages:

- GeneratorInterface/Pythia8Interface (**generators**)


@SiewYan, @mkirsano, @Saptaparna, @cmsbuild, @alberto-sanchez, @menglu21, @GurpreetSinghChahal can you please review it and eventually sign? Thanks.
@alberto-sanchez, @mkirsano this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

vslokenb committed:
Update TopRecoilHook.h

Fix suggested by PR 42313. Removes useless check.

cmsbuild commented:
Pull request #42332 was updated. @SiewYan, @mkirsano, @Saptaparna, @cmsbuild, @alberto-sanchez, @menglu21, @GurpreetSinghChahal can you please check and sign again.

perrotta commented:
backport of #42180

menglu21 commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-dfc0e9/34017/summary.html
**COMMIT**: f4e5c6c32bfb7cf287c90726888e4cc838961404
**CMSSW**: CMSSW_13_2_X_2023-08-01-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/42332/34017/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 4 lines to the logs
 - Reco comparison results: 13 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3195634
 - DQMHistoTests: Total failures: 9
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3195603
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_2_X_2023-08-01-1100+dfc0e9/58198/triggerResults): no differences found

perrotta commented on GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h:
```diff
@@ -0,0 +1,132 @@
+        partonSystemsPtr->list();
+        event.list();
+      }
+      //std::cout << "BEGIN TOP TEST: "  << (wtT / wtW) << " | END TOP TEST" << endl;
```
Please remove
```suggestion
```

perrotta commented on GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    // Destructor prints histogram.
+     ~TopRecoilHook() override { delete wtCorr; }
```
Please align
```suggestion
    ~TopRecoilHook() override { delete wtCorr; }
```

perrotta reviewed (commented):
This being based on the rather old CMSSW_3_2_0_pre2, it was a bit complicate to compare with the same PR merged in master.
At the end I managed to do it.: therefore, even if a rebase on the more recent 13_2_X IB could help, it is not mandatory, provided you apply the two small adjustments listed here needed to synchronize with the master version of the same PR.
After that (and the @cms-sw/generators-l2 signature, of course) this PR can be merged

vslokenb committed:
Update GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h

Co-authored-by: Andrea Perrotta <perrotta@cern.ch>

vslokenb committed:
Update GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h

Co-authored-by: Andrea Perrotta <perrotta@cern.ch>

cmsbuild commented:
Pull request #42332 was updated. @SiewYan, @mkirsano, @Saptaparna, @cmsbuild, @alberto-sanchez, @menglu21, @GurpreetSinghChahal can you please check and sign again.

perrotta commented on GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc:
```diff
@@ -29,7 +30,6 @@ using namespace Pythia8;
 // Emission Veto Hooks
 //
 #include "Pythia8Plugins/PowhegHooks.h"
-#include "Pythia8Plugins/PowhegHooksVincia.h"
```
This PR backports in 13_2_X also https://github.com/cms-sw/cmssw/pull/42020
However, that PR is supposed to go with https://github.com/cms-sw/cmsdist/pull/8552, which as far as I can see is not backported in 13_2.
Please backport in 13_2 only what is needed for this PR. Updates of Pythia in closed release cycles are delicate and should be either avoided or dealt with by @cms-sw/generators-l2 in a separate and dedicated PR

perrotta commented:
hold
(see https://github.com/cms-sw/cmssw/pull/42332#pullrequestreview-1591107324)

cmsbuild commented:
Pull request has been put on hold by  @perrotta
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-dfc0e9/34443/summary.html
**COMMIT**: 81dde92486885f44f184837655d7740fa27ae199
**CMSSW**: CMSSW_13_2_X_2023-08-22-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/42332/34443/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 12 lines to the logs
 - Reco comparison results: 6 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3196338
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3196313
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_2_X_2023-08-22-2300+dfc0e9/58561/triggerResults): no differences found

menglu21 commented:
@vslokenb kind ping

vslokenb committed:
Restore extra backport changes

PowhegHooksVincia removed because of potential error with PR #42020, however this commit exactly copies the file from the 13_1 backport. Hopefully this remedies the issue

cmsbuild commented:
Pull request #42332 was updated. @SiewYan, @mkirsano, @bbilin, @cmsbuild, @alberto-sanchez, @menglu21, @GurpreetSinghChahal can you please check and sign again.

vslokenb commented:
Thanks for the reminder, hopefully this fixes the issue!

perrotta commented:
unhold
(now this PR truly reflects what is in the backported PRs #42180 and #42313)

menglu21 commented on GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc:
```diff
@@ -19,6 +19,7 @@ using namespace Pythia8;
 
 #include "GeneratorInterface/Pythia8Interface/plugins/ReweightUserHooks.h"
```
in https://github.com/cms-sw/cmssw/pull/42180/, it's using "#include "ReweightUserHooks.h""

menglu21 commented on GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc:
```diff
@@ -19,6 +19,7 @@ using namespace Pythia8;
 
 #include "GeneratorInterface/Pythia8Interface/plugins/ReweightUserHooks.h"
 #include "GeneratorInterface/Pythia8Interface/interface/CustomHook.h"
+#include "GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h"
```
in https://github.com/cms-sw/cmssw/pull/42180, it's using "#include "TopRecoilHook.h"

menglu21 commented on GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    // Destructor prints histogram.
+    ~TopRecoilHook() override { delete wtCorr; }
```
in https://github.com/cms-sw/cmssw/pull/42180, it's:
```
 ~TopRecoilHook() override {
      if (doTopRecoil)
        ;
      delete wtCorr;
    }
```

perrotta replied:
Please don't revert this fix, which was merged in master with #42313 !
(Se also my comment https://github.com/cms-sw/cmssw/pull/42332#issuecomment-1755380830 above, when I unheld this PR)

vslokenb committed:
Update Pythia8Hadronizer.cc

cmsbuild commented:
Pull request #42332 was updated. @SiewYan, @mkirsano, @bbilin, @cmsbuild, @alberto-sanchez, @menglu21, @GurpreetSinghChahal can you please check and sign again.

vslokenb committed:
Update TopRecoilHook.h

cmsbuild commented:
Pull request #42332 was updated. @SiewYan, @mkirsano, @bbilin, @cmsbuild, @alberto-sanchez, @menglu21, @GurpreetSinghChahal can you please check and sign again.

menglu21 commented:
please test

menglu21 commented:
@cmsbuild please test

cmsbuild commented:
-1

**Failed Tests**: RelVals-INPUT 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-dfc0e9/34956/summary.html
**COMMIT**: cce687939364d682045095b1214e70774b7bdb91
**CMSSW**: CMSSW_13_2_X_2023-09-28-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/42332/34956/install.sh` to create a dev area with all the needed externals and cmssw changes.

Found compilation warnings

## RelVals-INPUT

The relvals timed out after 4 hours.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 5 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3198243
 - DQMHistoTests: Total failures: 4
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3198217
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_2_X_2023-09-28-1100+dfc0e9/59057/triggerResults): no differences found

menglu21 commented:
@perrotta is the error related to this PR

perrotta commented on GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h:
```diff
@@ -0,0 +1,135 @@
+    }
+
+    // Destructor prints histogram.
+     ~TopRecoilHook() override {
+      if (doTopRecoil)
+        ;
+      delete wtCorr;
+    }
```
```suggestion
    ~TopRecoilHook() override { delete wtCorr; }
```

vslokenb committed:
Update GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h

Co-authored-by: Andrea Perrotta <perrotta@cern.ch>

cmsbuild commented:
Pull request #42332 was updated. @GurpreetSinghChahal, @menglu21, @alberto-sanchez, @bbilin, @SiewYan, @cmsbuild, @mkirsano can you please check and sign again.

menglu21 commented:
@cmsbuild please test

menglu21 commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_13_2_X IBs (tests are also fine) and once validation in the development release cycle CMSSW_13_3_X is complete. This pull request will now be reviewed by the release team before it's merged. @rappoccio, @sextonkennedy, @antoniovilela (and backports should be raised in the release meeting by the corresponding L2)

rappoccio commented:
@perrotta are you okay with the PR now? Have your comments been addressed?

perrotta commented:
> @perrotta are you okay with the PR now? Have your comments been addressed?

Ciao Sal. Yes, now this PR correctly implements the fix that was merged in master with https://github.com/cms-sw/cmssw/pull/42313

antoniovilela commented:
+1
